### PR TITLE
[Continuous Batching] Optimizing top_p path for multinomial sample 

### PR DESCRIPTION
**Pull request by mzegla**
Summary of changes:
- Adding partial sorting for 10, 100 and 1000 elements. Since most of the time only a few logits carry most of the probability, we don't need all the logits sorted to apply `top_p`. 
- If 1000 top elements do no sum up to `top_p` we run regular sorting and consider the rest of the elements.

### CODE REVIEW

File: src/cpp/src/logit_processor.hpp [hunk 1]
```diff
@@ -58,22 +58,45 @@ class TopPFilter : public ILogitTransformer {
...
-            logits.initialize_vector();
-            std::sort(logits.m_vector.begin(), logits.m_vector.end(), [](const Token& lhs, const Token& rhs) {return lhs.m_log_prob > rhs.m_log_prob; });
+    bool partial_sort_and_resize(Logits& logits) {
+        // Since most of the time huge part of logits vector contains minimal values 
+        // expensive sorting of entire vector might be unnecessary, especially for low values of top_p. 
+        // This method partially sorts vector 3 times considering 10, 100 and 1000 top elements and stops when top_p condition is met.
+        // If top_p is found in considered scope it resizes logits vector and returns true. Otherwise it returns false and regular sorting is performed.
+        // Note that it can we less performant than standard approach if logits value are more evenly distributed across the vector.
+        float sum = 0.0;
+        for (uint32_t step = 10; step <= 1000; step *= 10) {
```
Comment (ilya-lavrenov) [inline src/cpp/src/logit_processor.hpp:68]: do you think that multiplier should be 2? 
E.g. start with 16/32 and multiple by 2.. I am not sure about real data, but it seems that difference between sorting 100 and 1000 can be too big.

**Review by ilya-lavrenov — APPROVED** [review]
